catalog/app/utils/spreadsheets/spreadsheets.spec.ts

```diff
@@ -0,0 +1,287 @@
+        a: ['b', 'c'],
+        d: ['e,i,j,k', 'f'],
+        g: ['h', null],
+      })
+    })
+  })
+
+  describe('scoreObjectDiff', () => {
+    it('object with more keys in commont has bigger score', () => {
+      const testA = { a: 1, b: 2, c: 3 }
+      const testB = { c: 4, d: 5, e: 6 }
+      const control = { b: 123, c: 456 }
+      const scoreA = spreadsheets.scoreObjectDiff(testA, control)
+      const scoreB = spreadsheets.scoreObjectDiff(testB, control)
+      expect(scoreA > scoreB).toBe(true)
+    })
+  })
+
+  describe('parseSpreadsheetAgainstSchema', () => {
+    const schema = {
+      type: 'object',
+      properties: {
+        b: { type: 'number' },
+        c: { type: 'number' },
+      },
+    }
+
+    it('parse vertical spreadsheet', () => {
+      const csv = dedent`
+        b,c
+        1,2
+      `
+      const workbook = xlsx.read(csv, { type: 'string', cellDates: true })
+      const sheet = workbook.Sheets[workbook.SheetNames[0]]
+      expect(spreadsheets.parseSpreadsheetAgainstSchema(sheet, schema)).toEqual({
+        b: [1],
+        c: [2],
+      })
+    })
+
+    it('parse horizontal spreadsheet', () => {
+      const csv = dedent`
+        b,1
+        c,2
+      `
+      const workbook = xlsx.read(csv, { type: 'string', cellDates: true })
+      const sheet = workbook.Sheets[workbook.SheetNames[0]]
+      expect(spreadsheets.parseSpreadsheetAgainstSchema(sheet, schema)).toEqual({
+        b: [1],
+        c: [2],
+      })
+    })
+
+    it('parse as vertical when no keys in common', () => {
+      const csv = dedent`
+        d,e,f
+        1,2,3
+      `
+      const workbook = xlsx.read(csv, { type: 'string', cellDates: true })
+      const sheet = workbook.Sheets[workbook.SheetNames[0]]
+      expect(spreadsheets.parseSpreadsheetAgainstSchema(sheet, schema)).toEqual({
+        d: [1],
+        e: [2],
+        f: [3],
+      })
+    })
+
+    it('parse invalid data with no error', () => {
+      const sheet = ['123']
+      expect(spreadsheets.parseSpreadsheetAgainstSchema(sheet, schema)).toEqual({})
+    })
+
+    describe('for flat Excel files', () => {
+      const bilboSchema = {
+        type: 'object',
+        properties: {
+          Fingers: { type: 'array', items: { type: 'array' } },
+          Male: { type: 'array', tems: { type: 'boolean' } },
+          Parts: { type: 'array', items: { type: 'array' } },
+        },
+      }
+      const outputRaw = {
+        Age: [131],
+        Date: ['1990-09-22'],
+        Fingers: ['1,2,3,4,5,6,7,8,9,10'],
+        Male: [true],
+        Name: ['Bilbo Beggins'],
+        Parts: ['head,legs,arms'],
+      }
+      const outputSchemed = {
+        Age: [131],
+        Date: ['1990-09-22'],
+        Fingers: [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]],
+        Male: [true],
+        Name: ['Bilbo Beggins'],
+        Parts: [['head', 'legs', 'arms']],
+      }
+
+      it('parses OpenOffice archive format', () => {
```

Comment:
i think you can extract all this repeated code into a function and reuse it

Reply:
I would like to do it in the next PR